utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,369 @@
+    RESULT_ENSURE(pthread_rwlock_rdlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+    *return_fork_generation_number = fgn_state.current_fork_generation_number;
+    if (*fgn_state.zero_on_fork_addr != S2N_FORK_EVENT) {
+        /* No fork event detection. */
+        RESULT_ENSURE(pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+        return S2N_RESULT_OK;
+    }
+    RESULT_ENSURE(pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+
+    /* We are mutating the process-global, cached fork generation number. Need
+     * to acquire the write lock for that. Set returned fgn before checking the
+     * if condition with the same reasons as above.
+     */
+    RESULT_ENSURE(pthread_rwlock_wrlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+    *return_fork_generation_number = fgn_state.current_fork_generation_number;
+    if (*fgn_state.zero_on_fork_addr == S2N_FORK_EVENT) {
+        /* Fork event has been detected; reset sentinel, increment cached fork
+         * generation nunber (which is now "current" in this child process), and
+         * write incremented fork generation number to the output parameter.
+         */
+        *fgn_state.zero_on_fork_addr = S2N_NO_FORK_EVENT;
+        fgn_state.current_fork_generation_number = fgn_state.current_fork_generation_number + 1;
+        *return_fork_generation_number = fgn_state.current_fork_generation_number;
+    }
+    RESULT_ENSURE(pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+
+    return S2N_RESULT_OK;
+}
+
+/* Run-time probe checking whether the system supports the MADV_WIPEONFORK fork
+ * detection mechanism.
+ *
+ * Return value:
+ *  If not supported, returns false.
+ *  If supported, returns true.
+ */
+static bool s2n_probe_madv_wipeonfork_support(void) {
+
+    bool result = false;
+
+#if defined(USE_MADVISE)
+    void *probe_addr = MAP_FAILED;
+    long page_size = 0;
+
+    page_size = sysconf(_SC_PAGESIZE);
+    if (page_size <= 0) {
+        return false;
+    }
+
+    probe_addr = mmap(NULL, (size_t) page_size, PROT_READ | PROT_WRITE,
+                MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (probe_addr == MAP_FAILED) {
+        return false;
+    }
+
+    /* Some versions of qemu (up to at least 5.0.0-rc4, see
+     * linux-user/syscall.c) ignore invalid advice arguments. Hence, we first
+     * verify that madvise() rejects advice arguments it doesn't know about.
+     */
+    if (madvise(probe_addr, (size_t) page_size, -1) != 0 &&
+        madvise(probe_addr, (size_t) page_size, MADV_WIPEONFORK) == 0) {
+        result = true;
+    }
+
+    munmap(probe_addr, (size_t) page_size);
+#endif
+
+    return result;
+}
+
+bool s2n_assert_madv_wipeonfork_is_supported(void)
+{
+    return s2n_probe_madv_wipeonfork_support();
+}
```

Comment:
What is the purpose of this method? If you follow my comment on s2n_probe_madv_wipeonfork_support, this could become:
```suggestion

bool s2n_assert_madv_wipeonfork_is_supported(void)
{
    return s2n_result_is_ok(s2n_probe_madv_wipeonfork_support());
}
```